### PR TITLE
stdlib: io: fix RawIOBase and BufferedIOBase not being BinaryIO

### PR DESCRIPTION
**Pull request by FFY00**
Both RawIOBase and BufferedIOBase are binary I/O streams.
3058bec873ef8b2b5630df36c4cbea0e3734dc5f put the BinaryIO superclass at
the wrong level, it put in in the RawIOBase and BufferedIOBase
subclasses instead :/

Currently, type checkers don't think that open(..., 'rb') files are
BufferedIOBase, and open(..., 'rb', buffering=False) are RawIOBase.

https://docs.python.org/3/library/io.html#class-hierarchy
https://docs.python.org/3/library/io.html#io.RawIOBase
https://docs.python.org/3/library/io.html#io.BufferedIOBase

Signed-off-by: Filipe Laíns <lains@riseup.net>

### CODE REVIEW

**Comment by FFY00** [comment]
Hum, I am not quite sure what to make of this...

```
/home/runner/work/typeshed/typeshed/stdlib/io.pyi
  /home/runner/work/typeshed/typeshed/stdlib/io.pyi:57:30 - error: Class cannot derive from itself
  /home/runner/work/typeshed/typeshed/stdlib/io.pyi:57:30 - error: Base class type is unknown, obscuring type of derived class (reportUntypedBaseClass)
```

File: stdlib/io.pyi [hunk 1]
```diff
@@ -48,13 +48,13 @@ class IOBase:
...
 
-class RawIOBase(IOBase):
+class RawIOBase(IOBase, BinaryIO):
     def readall(self) -> bytes: ...
     def readinto(self, __buffer: WriteableBuffer) -> int | None: ...
     def write(self, __b: ReadableBuffer) -> int | None: ...
     def read(self, __size: int = ...) -> bytes | None: ...
 
-class BufferedIOBase(IOBase):
+class BufferedIOBase(IOBase, BufferedIOBase):
```
Comment (Akuli) [inline stdlib/io.pyi:57]: I don't know what you are trying to do here, but `class Foo(Bar, Foo):` doesn't make any sense, and that's what pyright complains about.

Comment (FFY00) [reply stdlib/io.pyi]: Ugh, it seems my dyslexia yesterday was making me read a totally different word, sometimes my brain gets the visual mental cache wrong :sweat_smile:

This should of course should be `BinaryIO`.

**Comment by JelleZijlstra** [comment]
This still has a number of CI failures.

**Comment by JelleZijlstra** [comment]
Are you still interested in moving this forward?

Personally I don't think it's worth massaging the concrete IO classes too much. Use protocols instead.

**Comment by FFY00** [comment]
I am but I am not sure how to fix this.

> Personally I don't think it's worth massaging the concrete IO classes too much. Use protocols instead.

I disagree, `open` is type-hinted to return the concrete IO classes so mypy complains when we try to use the protocols that should be there.

**Comment by JelleZijlstra** [comment]
This change doesn't work as is and now has a merge conflict. If you're interested in pushing it to completion, feel free to open a new PR.

**Pull request closed without merging**